src/core/lib/iomgr/tcp_posix.cc

```diff
@@ -79,6 +90,175 @@ typedef size_t msg_iovlen_type;
 
 extern grpc_core::TraceFlag grpc_tcp_trace;
 
+namespace grpc_core {
+
+class TcpZcpSendRecord {
+ public:
+  struct OutgoingOffset {
+    size_t slice_idx = 0;
+    size_t byte_idx = 0;
+  };
+
+  TcpZcpSendRecord() { grpc_slice_buffer_init(&buf_); }
+  ~TcpZcpSendRecord() {
+    AssertEmpty();
+    grpc_slice_buffer_destroy_internal(&buf_);
+  }
+  OutgoingOffset& GetOutgoingOffsets() { return out_offset_; }
+  grpc_slice_buffer* Buffer() { return &buf_; }
+  void PrepareForSends(grpc_slice_buffer* slices_to_send) {
+    AssertEmpty();
+    out_offset_.slice_idx = 0;
+    out_offset_.byte_idx = 0;
+    grpc_slice_buffer_swap(slices_to_send, &buf_);
+    Ref();
+  }
+  void Ref() { ref_.FetchAdd(1, MemoryOrder::RELAXED); }
+  bool Unref() {
+    const intptr_t prior = ref_.FetchSub(1, MemoryOrder::ACQ_REL);
+    GPR_DEBUG_ASSERT(prior > 0);
+    if (prior == 1) {
+      AllSendsComplete();
+      return true;
+    }
+    return false;
+  }
+
+ private:
+  void AssertEmpty() {
+    GPR_DEBUG_ASSERT(buf_.count == 0);
+    GPR_DEBUG_ASSERT(buf_.length == 0);
+    GPR_DEBUG_ASSERT(ref_.Load(MemoryOrder::RELAXED) == 0);
+  }
+  void AllSendsComplete() {
+    GPR_DEBUG_ASSERT(ref_.Load(MemoryOrder::RELAXED) == 0);
+    grpc_slice_buffer_reset_and_unref_internal(&buf_);
+  }
+
+  grpc_slice_buffer buf_;
+  Atomic<intptr_t> ref_;
+  OutgoingOffset out_offset_;
+};
+
+class TcpZcpSendCtx {
+ public:
+  static constexpr int kDefaultMaxSends = 4;
+  TcpZcpSendCtx(int max_sends = kDefaultMaxSends)
+      : max_sends_(max_sends), free_send_records_size_(max_sends) {
+    send_records_ = static_cast<TcpZcpSendRecord*>(
+        gpr_malloc(max_sends * sizeof(*send_records_)));
+    free_send_records_ = static_cast<TcpZcpSendRecord**>(
+        gpr_malloc(max_sends * sizeof(*free_send_records_)));
+    if (!send_records_ || !free_send_records_) {
+      gpr_free(send_records_);
+      gpr_free(free_send_records_);
+      gpr_log(GPR_INFO, "Disabling TCP TX zerocopy due to memory pressure.\n");
+      memory_limited_ = true;
+    } else {
+      for (auto idx = 0; idx < max_sends_; ++idx) {
+        new (send_records_ + idx) TcpZcpSendRecord();
+        free_send_records_[idx] = send_records_ + idx;
+      }
+    }
+  }
+  ~TcpZcpSendCtx() {
+    if (send_records_) {
+      for (auto idx = 0; idx < max_sends_; ++idx) {
+        send_records_[idx].~TcpZcpSendRecord();
+      }
+    }
+    gpr_free(send_records_);
+    gpr_free(free_send_records_);
+  }
+  bool IsMemoryLimited() { return memory_limited_; }
```

Comment:
Non-blocking comment - This could also simply be memory_limited() since it is a getter function.